File: CSharp/DependencyBreakingKatas/D/RestCountriesAPI.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Org.Codecop.Dependencies.D.Restcountries;
using unirest_net.http;
using Newtonsoft.Json;

namespace Org.Codecop.Dependencies.D
{
    public class RestCountriesAPI
    {
        private static readonly Country HomeBase = new Country("AT");

        private const string CountryInformationServiceUrl = "https://api.restcountries.com/countries/v5?pretty=1&limit=100";

        private static readonly RestCountriesAPI instance = new RestCountriesAPI();

        private RestCountriesAPI()
        {
        }

        public static RestCountriesAPI GetInstance()
        {
            return instance;
        }

        public bool IsInCommonMarket(Country country)
        {
            CountryDescription countryDescription = GetCountryDescriptionViaRestCall(country);
            return countryDescription?.region.Equals("Europe") ?? false;
        }

        public bool IsInAmericas(Country country)
        {
            CountryDescription countryDescription = GetCountryDescriptionViaRestCall(country);
            return countryDescription?.region.Equals("Americas") ?? false;
        }

        public int DistanceTo(Country country)
        {
            CountryDescription austria = GetCountryDescriptionViaRestCall(HomeBase);
            CountryDescription other = GetCountryDescriptionViaRestCall(country);

            if (austria == null || other == null)
            {
                throw new RestCountriesAPIException("Could not find country " + HomeBase + " or " + country);
            }

            return DistBetween(austria.coordinates.lat, austria.coordinates.lng, 
                               other.coordinates.lat, other.coordinates.lng);
        }

        private int DistBetween(double fromLatitude, double fromLongitude, double toLatitude, double toLongitude)
        {
            double earthRadius = 6371000; // meters
            double fLat = ConvertToRadians(fromLatitude);
            double toLat = ConvertToRadians(toLatitude);
            double diffLat = ConvertToRadians(toLatitude - fromLatitude);
            double diffLng = ConvertToRadians(toLongitude - fromLongitude);
            double a = Math.Sin(diffLat / 2) * Math.Sin(diffLat / 2)
                + Math.Cos(fLat) * Math.Cos(toLat) * Math.Sin(diffLng / 2) * Math.Sin(diffLng / 2);
            double c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
            float dist = (float)(earthRadius * c);
            return (int)dist;
        }

        private double ConvertToRadians(double angle)
        {
            return (Math.PI / 180) * angle;
        }

        private CountryDescription GetCountryDescriptionViaRestCall(Country country)
        {
            return SlowHttpCall().
                Where(c => c.codes.alpha_2.Equals(country.ToString())).
                SingleOrDefault();
        }

        public IList<CountryDescription> SlowHttpCall()
        {
            List<CountryDescription> countryDescriptions = new List<CountryDescription> {};

            try
            {
                for (int offset = 0; offset < 300; offset += 100) {
                    System.Threading.Thread.Sleep(300);
                    var jsonResponse = Unirest //
                        .get(CountryInformationServiceUrl + "&offset=" + offset)
                        .header("accept", "application/json")
                        .header("Authorization", "Bearer " + System.Environment.GetEnvironmentVariable("RESTCOUNTRIES_API_KEY"))
                        .asString();

                    var body = jsonResponse.Body;
                    var jsonBody = JsonConvert.DeserializeObject<Root>(body);
                    countryDescriptions.AddRange(jsonBody.data.objects);
                }

            }
            catch (Exception e)
            {
                throw new RestCountriesAPIException("Could not read country information from " + CountryInformationServiceUrl, e);
            }

            return countryDescriptions;
        }
        
        public static void Main(string[] args)
        {
            Console.WriteLine(new Country("US"));
            Console.WriteLine(GetInstance().IsInAmericas(new Country("US")));
            Console.WriteLine(GetInstance().IsInCommonMarket(new Country("US")));
            Console.WriteLine(GetInstance().DistanceTo(new Country("US")));
            Console.WriteLine(new Country("AT"));
            Console.WriteLine(GetInstance().IsInAmericas(new Country("AT")));
            Console.WriteLine(GetInstance().IsInCommonMarket(new Country("AT")));
            Console.WriteLine(GetInstance().DistanceTo(new Country("AT")));
        }
    }
}

File: PHP/src/D/RestCountriesAPI.php
<?php

namespace Org\Codecop\Dependencies\D;

use Unirest\Request;

class RestCountriesAPI {

    private const HOME_BASE = 'AT';
    private const COUNTRY_INFORMATION_SERVICE_URL = 'https://api.restcountries.com/countries/v5?pretty=1&limit=100';

    private static $instance;

    public static function getInstance() {
        if (!isset(self::$instance)) {
            self::$instance = new RestCountriesAPI();
        }
        return self::$instance;
    }

    public function isInCommonMarket(Country $country): bool {
        $countryDescription = $this->getCountryDescriptionViaRestCall($country);
        return $countryDescription !== null && $countryDescription->region === 'Europe';
    }

    public function isInAmericas(Country $country): bool {
        $countryDescription = $this->getCountryDescriptionViaRestCall($country);
        return $countryDescription !== null && $countryDescription->region === 'Americas';
    }

    public function distanceTo(Country $country): int {
        $austria = $this->getCountryDescriptionViaRestCall(new Country(self::HOME_BASE));
        $other = $this->getCountryDescriptionViaRestCall($country);

        if ($austria === null || $other === null) {
            throw new RestCountriesAPIException('Could not find country ' . self::HOME_BASE . ' or ' . $country);
        }

        return $this->distBetween(
            $austria->coordinates->lat, $austria->coordinates->lng,
            $other->coordinates->lat, $other->coordinates->lng,
        );
    }

    private function distBetween(float $fromLatitude, float $fromLongitude, float $toLatitude, float $toLongitude): int {
        $earthRadius = 6371000; // meters
        $fLat = deg2rad($fromLatitude);
        $toLat = deg2rad($toLatitude);
        $diffLat = deg2rad($toLatitude - $fromLatitude);
        $diffLng = deg2rad($toLongitude - $fromLongitude);
        $a = sin($diffLat / 2) * sin($diffLat / 2)
            + cos($fLat) * cos($toLat) * sin($diffLng / 2) * sin($diffLng / 2);
        $c = 2 * atan2(sqrt($a), sqrt(1 - $a));
        $dist = $earthRadius * $c;
        return (int) $dist;
    }

    private function getCountryDescriptionViaRestCall(Country $country) {
        $countryDescriptions = $this->slowHttpCall();
        foreach ($countryDescriptions as $c) {
            if ($c->codes->alpha_2 === $country->__toString()) {
                return $c;
            }
        }
        return null;
    }

    public function slowHttpCall(): array {
        $countryDescriptions = [];

        try {
            for ($offset = 0; $offset < 300; $offset += 100) {
                usleep(300000); // Sleep for 1 second (in microseconds)
                $headers = [
                    'accept'  => 'application/json',
                    'Authorization' => 'Bearer ' . getenv('RESTCOUNTRIES_API_KEY')
                ];
                $response = Request::get(self::COUNTRY_INFORMATION_SERVICE_URL . "&offset=" . $offset, $headers);
                $jsonBody = json_decode($response->raw_body);
                $countryDescriptions = array_merge($countryDescriptions, $jsonBody->data->objects);
            }

        } catch (\Exception $e) {
            throw new RestCountriesAPIException('Could not read country information from ' . self::COUNTRY_INFORMATION_SERVICE_URL, 0, $e);
        }

        return $countryDescriptions;
    }

}
